### ISSUE TITLE
Strategy Optimisation

### ISSUE THREAD

**Issue by karunkrishna** [karunkrishna/pythonAlgo#13]
We are going to describe optimisation methods to improve the performance of our trading strategies by tuning the parameters in a systematic fashion.

For this we will use mechanisms from the statistical field of Model Selection, such as cross-validation and grid search

In momentum strategies using technical indicators, such as with moving averages (simple or exponential), there is a need to specify a lookback window. The same is true of many mean-reverting strategies, which require a (rolling) lookback window in order to calculate a regression between two time series. Particular statistical machine learning models such as a logistic regression, SVM or Random Forest also require parameters in order to be calculated.

The biggest danger when considering parameter optimisation is that of overfitting a model or trading strategy. This problem occurs when a model is trained on an in sample retained slice of training data and is optimised to perform well (by the appropriate performance measure), but performance degrades substantially when applied to out of sample data. For instance, a trading strategy could perform extremely well in the backtest (the in sample data) but when deployed for live trading can be completely unprofitable.

An additional concern of parameter optimisation is that it can become very computationally expensive. With modern computational systems this is less of an issue than it once was, due to parallelisation and fast CPUs. However, multiple parameter optimisation can increase computational complexity by orders of magnitudes. One must be aware of this as part of the research and development process.

In addition to parameters there are numerous means of evaluating the performance of a statistical model and the trading strategy based upon it. We have defined concepts such as the hit rate and the confusion matrix. In addition there are more statistical measures such as the Mean Squared Error (MSE). These are performance measures that would be optimised at the statistical model level, via parameters relevant to their domain.

The actual trading strategy is evaluated on different criteria, such as compound annual growth rate (CAGR) and maximum drawdown. We would need to vary entry and exit criteria, as well as other thresholds that are not directly related to the statistical model. Hence this motivates the question as to which set of parameters to optimise and when.

**Comment by karunkrishna** [comment]
**Optimisation is Expensive**

With multiple real-valued parameters, optimisation can rapidly become extremely expensive, as each new parameter adds an additional spatial dimension. 

![image](https://cloud.githubusercontent.com/assets/2157606/24017151/3f64e2bc-0a65-11e7-9a83-519069925bfe.png)

**Comment by karunkrishna** [comment]
**Overfitting**

Overfitting is the process of optimising a parameter, or set of parameters, against a particular data set such that an appropriate performance measure (or error measure) is found to be maximised (or minimised), but when applied to an unseen data set, such a performance measure degrades substantially. The concept is closely related to the idea of the bias-variance dilemma. 

The bias-variance dilemma concerns the situation where a statistical model has a trade-off between being a low-bias model or a low-variance model, or a compromise between the two. Bias refers to the difference between the model’s estimation of a parameter and the true "population" value of the parameter, or erroneous assumptions in the statistical model. Variance refers to the error from the sensitivity of the model to small fluctuations in the training set (in sample data). 

In all statistical models one is simultaneously trying to minimise both the bias error and the variance error in order to improve model accuracy. Such a situation can lead to overfitting in models, as the training error can be substantially reduced by introducing models with more flexibility (variation). However, such models can perform extremely poorly on new (out of sample) data since they were essentially "fit" to the in sample data. A common example of a high-bias, low-variance model is that of linear

**Comment by karunkrishna** [comment]
**Cross Validation**

Cross Validation is a technique used to assess how a statistical model will generalise to new data that it has not been exposed to before. Such a technique is usually used on predictive models, such as the aforementioned supervised classifiers used to predict the sign of the following daily returns of an asset price series. Fundamentally, the goal of cross validation is to minimise error on out of sample data without leading to an overfit model. 

In this section we will describe the training/test split and k-fold cross validation, as well as use techniques within Scikit-Learn to automatically carry out these procedures on statistical models we have already developed

**Comment by karunkrishna** [comment]
```


# Create the (parametrised) models
print("Hit Rates/Confusion Matrices:\n")
models = [("LR", LogisticRegression()),
	("LDA", LDA()),
	("QDA", QDA()),
	("LSVC", LinearSVC()),
	("RSVM", SVC(
	C=1000000.0, cache_size=200, class_weight=None,
	coef0=0.0, degree=3, gamma=0.0001, kernel=’rbf’,
	max_iter=-1, probability=False, random_state=None,
	shrinking=True, tol=0.001, verbose=False)),
	("RF", RandomForestClassifier(
	n_estimators=1000, criterion=’gini’,
	max_depth=None, min_samples_split=2,
	min_samples_leaf=1, max_features=’auto’,
	bootstrap=True, oob_score=False, n_jobs=1,
	random_state=None, verbose=0)
	)]

# Iterate through the models
for m in models:
	# Train each of the models on the training set
	m[1].fit(X_train, y_train)
	
	# Make an array of predictions on the test set
	pred = m[1].predict(X_test)
	
	# Output the hit-rate and the confusion matrix for each model
	print("%s:\n%0.3f" % (m[0], m[1].score(X_test, y_test)))
	print("%s\n" % confusion_matrix(pred, y_test))
```

**Comment by karunkrishna** [comment]
Consequently we can likely conclude that the particular choice of training/test split lead to an over-optimistic view of the predictive capability of the classifier. The next step is to increase the number of times a cross-validation is performed in order to minimise any potential overfitting. For this we will use k-fold cross validation.

K-Fold Cross Validation

![image](https://cloud.githubusercontent.com/assets/2157606/24025459/d89d2cc0-0a90-11e7-8cfd-a742a171c4fe.png)

**Comment by karunkrishna** [comment]
Grid Search

We now turn our attention to optimising the hyper-parameters of a particular statistical model. In essence they are the parameters that we need to specify when calling the initialisation of each statistical model. For this procedure we will use a process known as a grid search.

The basic idea is to take a range of parameters and assess the performance of the statistical model on each parameter element within the range. To achieve this in Scikit-Learn we can create a ParameterGrid. Such an object will produce a list of Python dictionaries that each contain a parameter combination to be fed into a statistical model.

**Comment by karunkrishna** [comment]
The next task is to modify the simulate_trading method to loop over all variants of strategy parameters. The method creates an output CSV file that is used to store parameter combinations and their particular performance metrics. This will allow us later to plot performance characteristics across parameters. 

The method loops over all of the strategy parameters and generates a new trading instance on every simulation. The backtest is then executed and the statistics calculated. These are stored and output into the CSV file. Once the simulation ends, the output file is closed:

**Comment by karunkrishna** [comment]
**Visualisation** 

As a final task in strategy optimisation, we are now going to visualise the performance characteristics of the backtester using Matplotlib, which is an extremely useful step when carrying out initial strategy research. Unfortunately we are the situation where we have a three-dimensional problem and so performance visualisation is not straightforward! However, there are some partial remedies to the situation. 

Firstly, we could fix the value of one parameter and take a "parameter slice" through the remainder of the "data cube". For instance we could fix the lookback window to be 100 and then see who the variation in z-score entry and exit thresholds affects the Sharpe Ratio or the maximum drawdown.

To achieve this we will use Matplotlib. We will read the output CSV and reshape the data such that we can visualise the results.
